### PR TITLE
drivers: sensor: bmm150: Add I2C-base or SPI-base interface in build …

### PR DESCRIPTION
**Pull request by Cherish-Gww**
…time

move DT_DRV_COMPAT to bmm150.h. so that can be decide which interface to use. define struct bmm150_bus_io interface for bmm150_i2c.c and bmm150_spi.c in bmm150.h. redefined bus operation interface in bmm150.c, this allow the driver to decide which interface to use during construction.

Signed-off-by: Weiwei Guo <guoweiwei@syriusrobotics.com>

### CODE REVIEW

File: drivers/sensor/bmm150/bmm150.c [hunk 1]
```diff
@@ -590,15 +618,38 @@ static int bmm150_init(const struct device *dev)
...
+ * Main instantiation macro, which selects the correct bus-specific
+ * instantiation macros for the instance.
+ */
+#define BMM150_DEFINE(inst)						\
+	static struct bmm150_data bmm150_data_##inst;			\
+	static const struct bmm150_config bmm150_config_##inst =	\
+		COND_CODE_1(DT_INST_ON_BUS(inst, spi),			\
+			    (BMM150_CONFIG_SPI(inst)),			\
+			    (BMM150_CONFIG_I2C(inst)));			\
+	DEVICE_DT_INST_DEFINE(inst,					\
```
Comment (MaureenHelm) [inline drivers/sensor/bmm150/bmm150.c:646]:
```suggestion
	SENSOR_DEVICE_DT_INST_DEFINE(inst,					\
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/bmm150/bmm150.c [hunk 2]
```diff
@@ -475,6 +505,9 @@ static int bmm150_attr_set(const struct device *dev,
 static const struct sensor_driver_api bmm150_api_funcs = {
 #if defined(BMM150_SET_ATTR)
 	.attr_set = bmm150_attr_set,
+#endif
+#ifdef CONFIG_BMM150_TRIGGER
+	.trigger_set = bmm150_trigger_set,
```
Comment (MaureenHelm) [inline drivers/sensor/bmm150/bmm150.c:510]: This change looks unrelated to SPI/I2C

File: drivers/sensor/bmm150/bmm150.c [hunk 3]
```diff
@@ -483,28 +516,27 @@ static const struct sensor_driver_api bmm150_api_funcs = {
...
-	if (bmm150_set_power_mode(dev, BMM150_POWER_MODE_SUSPEND, 1) < 0) {
-		LOG_ERR("failed to bring up device in suspend mode");
-		return -EIO;
+	/* Soft reset chip */
+	if (bmm150_reg_update_byte(dev, BMM150_REG_POWER, BMM150_MASK_SOFT_RESET,
+				   BMM150_SOFT_RESET) < 0) {
+		LOG_ERR("failed reset chip");
+		goto err_poweroff;
 	}
+	k_busy_wait(1000);
```
Comment (MaureenHelm) [inline drivers/sensor/bmm150/bmm150.c:528]: Please document why this particular wait duration was chosen. Can you sleep instead of busy-wait?

File: drivers/sensor/bmm150/bmm150.h [hunk 4]
```diff
@@ -114,11 +152,39 @@ struct bmm150_trim_regs {
...
+#endif
+
+#if defined(CONFIG_BMM150_TRIGGER_GLOBAL_THREAD) || \
+	defined(CONFIG_BMM150_TRIGGER_DIRECT)
+	const struct device *dev;
+#endif
+
+#ifdef CONFIG_BMM150_TRIGGER
+	sensor_trigger_handler_t handler_drdy;
+#endif /* CONFIG_BMM150_TRIGGER */
```
Comment (MaureenHelm) [inline drivers/sensor/bmm150/bmm150.h:187]: These changes look unrelated to SPI/I2C

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please move the trigger related changes to a separate commit

File: drivers/sensor/bmm150/bmm150.c (hunk 2 above)
Comment (Cherish-Gww) [reply drivers/sensor/bmm150/bmm150.c:510]: > This change looks unrelated to SPI/I2C

I will fix it, and take another PR for trigger mode.

File: drivers/sensor/bmm150/bmm150.c (hunk 3 above)
Comment (Cherish-Gww) [reply drivers/sensor/bmm150/bmm150.c:530]: > Please document why this particular wait duration was chosen. Can you sleep instead of busy-wait?

wait 1ms is comes from here:
https://github.com/boschsensortec/BMM150-Sensor-API/blob/a20641f216057f0c54de115fe81b57368e119c01/bmm150.c#L677

Comment (Cherish-Gww) [reply drivers/sensor/bmm150/bmm150.c]: > Please document why this particular wait duration was chosen. Can you sleep instead of busy-wait?

From suspend mode to sleep mode, it need sleep for 3ms, it comes here:
https://github.com/boschsensortec/BMM150-Sensor-API/blob/a20641f216057f0c54de115fe81b57368e119c01/bmm150.c#L1287

**Review by MaureenHelm — COMMENTED** [review]
The code changes look fine, but you need to address the failed compliance check:
```
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

 1: UC2 Signed-off-by: must have a full name
3: UC4 Commit body line exceeds max length (77>75): "move DT_DRV_COMPAT to bmm150.h. so that can be decide which interface to use."
4: UC4 Commit body line exceeds max length (84>75): "define struct bmm150_bus_io interface for bmm150_i2c.c and bmm150_spi.c in bmm150.h."
5: UC4 Commit body line exceeds max length (84>75): "redefined bus operation interface in bmm150.c, this allow the driver to decide which"
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

 7ed19f5ff6b1ff67d850d4b28bd71185[26](https://github.com/zephyrproject-rtos/zephyr/actions/runs/3871771124/jobs/6607174828#step:9:27)1ed350: author email (Cherish <guoweiwei@syriusrobotics.com>) does not follow the syntax: First Last <email>.
```

**Comment by MaureenHelm** [comment]
Close and reopen to kick CI

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
Needs rebase

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
@Cherish-Gww please rebase

**Comment by Cherish-Gww** [comment]
@carlescufi  @MaureenHelm  Rebase done.

**Review by teburd — COMMENTED** [review]
Is the bosch,bmm150 binding still needed here? Looks good to me, will await others for a +1 from myself.

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]
